# Pandas Module 1 Practice

### Dataset
Titanic CSV

### Goal
Revise core Pandas operations by actually running them on the Titanic dataset.

### My learning method
For each important operation I will note:

**Objective → Concept → Code → Output → My Observation → What I Learned**

I am keeping this notebook as a learning record, not as a perfect notebook.


# 1. Load the Titanic CSV

### Objective
Read the Titanic CSV file and store it in a DataFrame.

### Concept
`pd.read_csv()` loads CSV data into a Pandas DataFrame.


In [ ]:
import pandas as pd

df = pd.read_csv("../01_Data/Titanic.csv")

### Output
There is no printed output because I only loaded the data into `df(DataFrame)`.

### My Observation
The data is now stored in `df`, so I can use Pandas operations on it.

### What I Learned
`read_csv()` is the starting point when the data is available as a CSV file.


# 2. `head()` — Look at the Beginning

### Objective
See the first few records and understand what the dataset looks like.

### Concept
`df.head()` shows the first 5 rows by default.

In [2]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Output
The first 5 passenger records are displayed.

### My Observation
I can immediately see the column names such as `PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, and `Fare`. I can also see `NaN` values in places such as `Cabin`.

### What I Learned
`head()` is a quick way to inspect raw data before writing more analysis code.


### Practice: Show the First 10 Rows

In [3]:
df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


### My Observation
Passing `10` changes the number of rows displayed. The DataFrame itself is not changed.

### What I Learned
`head(n)` means "show the first `n` rows."


# 3. `tail()` — Look at the End

### Objective
Check the last records of the dataset.

### Concept
`df.tail()` returns the last 5 rows by default.

In [4]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


### Output
The final 5 rows are displayed.

### My Observation
The last record has index `890`, while its `PassengerId` is `891`.

### What I Learned
Pandas uses zero-based row positions, so row index and a dataset's ID column are not necessarily the same thing.


### Practice: Last 15 Rows

In [5]:
df.tail(15)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
876,877,0,3,"Gustafsson, Mr. Alfred Ossian",male,20.0,0,0,7534,9.8458,NaN,S
877,878,0,3,"Petroff, Mr. Nedelio",male,19.0,0,0,349212,7.8958,NaN,S
878,879,0,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C
880,881,1,2,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0,1,230433,26.0000,NaN,S
881,882,0,3,"Markun, Mr. Johann",male,33.0,0,0,349257,7.8958,NaN,S
882,883,0,3,"Dahlberg, Miss. Gerda Ulrika",female,22.0,0,0,7552,10.5167,NaN,S
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q


### My Observation
I get the final 15 records instead of only 5.

### What I Learned
`tail(n)` is useful when I want to inspect a specific number of records from the end.


# 4. `shape` — How Large Is My Dataset?

### Objective
Find the number of rows and columns.

### Concept
`df.shape` returns `(rows, columns)`.

In [6]:
df.shape

(891, 12)

### Output
```text
(891, 12)
```

### My Observation
The Titanic dataset has 891 rows and 12 columns.

### What I Learned
`shape` gives me a fast answer to a basic data question: "How big is this dataset?"


# 5. `columns` — Know My Variables

### Objective
See all column names available for analysis.

### Concept
`df.columns` returns the column labels.

In [7]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### Output
The main columns are:

`PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Cabin`, `Embarked`

### My Observation
Before filtering or selecting anything, I need to know the exact column names.

### What I Learned
Exact column names matter because Pandas uses them in expressions such as `df["Age"]`.


### Practice: Count Columns

In [8]:
len(df.columns)

12

### Output
```text
12
```

### My Observation
There are 12 columns.

### What I Learned
`len(df.columns)` can be used when I need the count instead of the names.


# 6. `info()` — Understand Structure

### Objective
Inspect data types, non-null counts, and the overall DataFrame structure.

### Concept
`df.info()` tells me how many entries and columns I have, how many values are non-null, and what dtype each column uses.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Output
Important parts of my actual output:

```text
891 entries, 0 to 890
PassengerId  891 non-null  int64
Survived     891 non-null  int64
Pclass       891 non-null  int64
Name         891 non-null  object
Sex          891 non-null  object
Age          714 non-null  float64
SibSp        891 non-null  int64
Parch        891 non-null  int64
Ticket       891 non-null  object
Fare         891 non-null  float64
Cabin        204 non-null  object
Embarked     889 non-null  object
```

### My Observation
`Age`, `Cabin`, and `Embarked` have missing values.

The numeric columns are:

`PassengerId`, `Survived`, `Pclass`, `Age`, `SibSp`, `Parch`, `Fare`

### What I Learned
`info()` is more than something I run for the sake of it. I should read its output and use it to decide what needs cleaning later.


# 7. `describe()` — Get a Statistical Summary

### Objective
Look at basic statistics for the numeric columns.

### Concept
`df.describe()` gives count, mean, standard deviation, minimum, quartiles, and maximum.

In [10]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Output
Some values from my actual run:

```text
Age mean      ≈ 29.699
Age median    = 28
Age maximum   = 80
Fare mean     ≈ 32.204
Fare maximum  = 512.3292
```

### My Observation
`describe()` helps me understand the range and central values of numeric data. I also notice that `Age` has a count of 714 instead of 891, which agrees with the missing-value information from `info()`.

### What I Learned
I should compare `count`, `mean`, `50%`, and `max` instead of looking only at one number.


# 8. Single Column Selection

### Objective
Work with only one column.

### Concept
`df["ColumnName"]` returns that column as a Pandas Series.

In [11]:
df["Age"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

### Output
The Age values for all passengers are displayed, including `NaN` where Age is missing.

### My Observation
Only one variable is returned instead of the whole DataFrame.

### What I Learned
A single-column selection is useful before calculations such as:

```python
df["Age"].mean()
```


# 9. Multiple Column Selection

### Objective
Keep only the columns needed for a particular question.

### Concept
Use a list of column names inside double brackets.

### Code


In [12]:
df[["Name", "Pclass", "Age"]]

,Name,Pclass,Age
0,"Braund, Mr. Owen Harris",3,22.0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0
2,"Heikkinen, Miss. Laina",3,26.0
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0
4,"Allen, Mr. William Henry",3,35.0
...,...,...,...
886,"Montvila, Rev. Juozas",2,27.0
887,"Graham, Miss. Margaret Edith",1,19.0
888,"Johnston, Miss. Catherine Helen ""Carrie""",3,NaN
889,"Behr, Mr. Karl Howell",1,26.0


### Output
A DataFrame with `Name`, `Pclass`, and `Age`.

### My Observation
The number of rows stays the same, but unnecessary columns are removed from the result.

### What I Learned
When a task only needs a few variables, selecting those columns makes the result easier to work with.


# 10. `iloc` — Select Rows by Position Index

### Objective
Access a row using its integer position(index).

### Concept
`iloc` means integer-location based indexing.

In [13]:
df.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

### Output
The first passenger's complete record is returned.

### My Observation
`df.iloc[0]` gives the first row because Pandas uses zero-based indexing.

### What I Learned
`iloc[]` is useful when I know the row position rather than a column condition.


### Practice: First 10 Rows with `iloc`

In [14]:
df.iloc[:10]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


### My Observation
This returns positions 0 through 9, which is 10 rows.

### What I Learned
The stop value in a Python slice is exclusive.


# 11. Homework Check — First 8 Rows

### Objective
Use `head()` to solve a specific homework instruction.

### Task
Show the first 8 rows.

### Code I wrote in homework


In [15]:
df.head(9)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S


### Output
The code displays **9 rows**, not 8.

### My Observation
I used `9` inside `head()`, so Pandas correctly returned the first 9 rows. The issue is not the function; it is my argument.

### What I Learned
When a question gives an exact number, I should check that the number inside the function matches the requirement.


# 12. Filtering — Female Passengers

### Objective
Return only female passengers.

### Concept
A Boolean condition can be used to filter rows.

In [16]:
df[df["Sex"] == "female"]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C
...,...,...,...,...,...,...,...,...,...,...,...,...
880,881,1,2,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0,1,230433,26.0000,NaN,S
882,883,0,3,"Dahlberg, Miss. Gerda Ulrika",female,22.0,0,0,7552,10.5167,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S


### Output
My actual homework output contains **314 rows**.

### My Observation
Every returned row has `Sex = female`. Male rows are removed.

### What I Learned
The general pattern is:

```python
df[condition]
```

I can reuse this pattern for many real data questions.


# 13. Filtering — Age Greater Than 18

### Objective
Find passengers whose Age is greater than 18.

### Concept
Comparison operators can create a Boolean condition for numeric data.

In [17]:
df[df["Age"] > 18]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


### Output
My actual homework output contains **575 rows**.

### My Observation
Rows with Age greater than 18 are kept. Rows with Age 18 or below are excluded. Missing Age values do not satisfy the condition.

### What I Learned
I can turn a sentence into a Pandas condition:

"Age greater than 18" → `df["Age"] > 18`


# 14. Sorting — Highest Fare First

### Objective
Sort passengers from the highest Fare to the lowest Fare.

### Concept
`sort_values()` sorts rows using a selected column.

In [18]:
df.sort_values("Fare", ascending=False)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
679,680,1,1,"Cardeza, Mr. Thomas Drake Martinez",male,36.0,0,1,PC 17755,512.3292,B51 B53 B55,C
258,259,1,1,"Ward, Miss. Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C
737,738,1,1,"Lesurer, Mr. Gustave J",male,35.0,0,0,PC 17755,512.3292,B101,C
88,89,1,1,"Fortune, Miss. Mabel Helen",female,23.0,3,2,19950,263.0000,C23 C25 C27,S
438,439,0,1,"Fortune, Mr. Mark",male,64.0,1,4,19950,263.0000,C23 C25 C27,S
...,...,...,...,...,...,...,...,...,...,...,...,...
806,807,0,1,"Andrews, Mr. Thomas Jr",male,39.0,0,0,112050,0.0000,A36,S
815,816,0,1,"Fry, Mr. Richard",male,NaN,0,0,112058,0.0000,B102,S
466,467,0,2,"Campbell, Mr. William",male,NaN,0,0,239853,0.0000,NaN,S
481,482,0,2,"Frost, Mr. Anthony Wood ""Archie""",male,NaN,0,0,239854,0.0000,NaN,S


### Output
The highest fare appears first. The top fare in this dataset is **512.3292**.

### My Observation
The order of rows changes according to Fare. The original `df` is not permanently replaced just because I called `sort_values()`.

### What I Learned
Sorting is useful when the task asks for top or bottom records.


# 15. Combined Practice — Like Internship Thinking

### Objective
Combine the operations I learned instead of treating each function as an isolated command.

### Task
Find female passengers older than 18 and sort them by Fare from highest to lowest.

In [19]:
df[(df["Sex"] == "female") & (df["Age"] > 18)].sort_values("Fare", ascending=False)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
258,259,1,1,"Ward, Miss. Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C
88,89,1,1,"Fortune, Miss. Mabel Helen",female,23.0,3,2,19950,263.0000,C23 C25 C27,S
341,342,1,1,"Fortune, Miss. Alice Elizabeth",female,24.0,3,2,19950,263.0000,C23 C25 C27,S
742,743,1,1,"Ryerson, Miss. Susan Parker ""Suzette""",female,21.0,2,2,PC 17608,262.3750,B57 B59 B63 B66,C
299,300,1,1,"Baxter, Mrs. James (Helene DeLaudeniere Chaput)",female,50.0,0,1,PC 17558,247.5208,B58 B60,C
...,...,...,...,...,...,...,...,...,...,...,...,...
276,277,0,3,"Lindblom, Miss. Augusta Charlotta",female,45.0,0,0,347073,7.7500,NaN,S
501,502,0,3,"Canavan, Miss. Mary",female,21.0,0,0,364846,7.7500,NaN,Q
106,107,1,3,"Salkjelsvik, Miss. Anna Kristine",female,21.0,0,0,343120,7.6500,NaN,S
649,650,1,3,"Stanley, Miss. Amy Zillah Elsie",female,23.0,0,0,CA. 2314,7.5500,NaN,S


### Output
A DataFrame containing only female passengers older than 18, ordered by Fare from high to low.

### My Observation
I can read this code from left to right:

1. `df["Sex"] == "female"` keeps female passengers.
2. `df["Age"] > 18` keeps passengers older than 18.
3. `&` means both conditions must be true.
4. `sort_values(..., ascending=False)` puts the highest Fare first.

### What I Learned
The important skill is not memorizing separate functions. It is combining small operations to answer a real question.


# Final Reflection — How I Should Observe My Code

I do not want to finish a cell by only looking at whether it ran without an error.

After every important code cell, I should ask:

### 1. What did I give Pandas?
Example:

```python
df["Age"]
```

I gave Pandas a request for one column.

### 2. What operation happened?
Example:

```python
df[df["Sex"] == "female"]
```

Pandas checked the Sex value row by row and kept the rows matching the condition.

### 3. What changed in the output?
For the female filter:

- Original rows = 891
- Result rows = 314

That change tells me the filter worked as intended.

### 4. What does the result tell me about the data?
Example:

```python
df.info()
```

I noticed missing values in Age, Cabin, and Embarked.

That is not just an output detail. It tells me something about the quality of the dataset.

### 5. What can I reuse later?
For example:

```python
df[df["Age"] > 18]
```

This teaches a reusable pattern:

**DataFrame + Boolean condition = filtered rows**

---

# My Observation Formula

For future Pandas work, I will use:

**Input → Operation → Output → Change → Meaning**

Example:

```python
df[df["Age"] > 18]
```

**Input:** Titanic DataFrame  
**Operation:** `Age > 18`  
**Output:** Matching passenger rows  
**Change:** Fewer rows than the original dataset  
**Meaning:** I isolated passengers whose recorded Age is greater than 18

---

# My Learning Rule

I should write observations in my own words.

Weak:
> `shape` returned `(891, 12)`.

Better:
> The dataset contains 891 rows and 12 columns.

Best:
> The dataset contains 891 passenger records and 12 variables, so I now know the size of the data I am working with.

That last sentence shows that I understood the output instead of only copying it.
